### ISSUE TITLE
“Memory Leaking” problem and temp solution

### ISSUE THREAD

**Issue by IridescentJiang** [kcyt/IntegratedPIFu#12]
Great project! I have successfully trained and obtained good results!

While running the training, I noticed that the server's memory usage keeps increasing within one epoch, reaching 100GB to 200GB, until it eventually throws an error due to high memory consumption. If memory usage is not excessively high, the next epoch releases the memory. 

I initially suspected it to be a memory leak issue and later came across this particular issue in the pytorch project: [https://github.com/pytorch/pytorch/issues/13246](https://github.com/pytorch/pytorch/issues/13246)

It informed me that it might not be an actual memory leak in the strict sense but rather a high memory consumption between multiple tasks caused by using Python's list or dictionary formats. 

Later, by examining the code, I pinpointed the problem to this specific line of code in 100-line of [TrainDataset.py](https://github.com/kcyt/IntegratedPIFu/blob/main/lib/data/TrainDataset.py). 
`self.mesh_dic = load_trimesh(self.mesh_directory,  training_subject_list = self.training_subject_list)  # a dict containing the meshes of all the CAD models.
`
I believe the replication of mesh_dic during getitem in multiple tasks is causing excessive memory usage. 

As a temporary solution, I moved the load mesh operation inside the select_sampling_method function, which resolved the high memory consumption issue. 
`mesh = trimesh.load(os.path.join(self.mesh_directory, subject, '%s.obj' % subject))`
However, this led to a new problem as data loading now takes much longer since load mesh needs to be performed for each getitem operation. 

I would like to ask if anyone has encountered a similar high memory consumption issue and how you resolved it.

My PyTorch version is 1.7.0 cu=11.0.

**Comment by IridescentJiang** [comment]
Thank you very much for the advice provided by the author. 

The author advised me to sample a large number of the sample points for each training mesh and store them in .npy files. Then, load the sample points and their labels while training and take subsets of those sample points and labels. 

I will try this suggestion and update the running results of the code modifications here.

**Comment by Yiju1213** [comment]
Yes, I also encountered the same issue.

The `__init__` of my Dataset also contains a list storing all the point clouds.

I quickly came across https://github.com/pytorch/pytorch/issues/13246 as you did, and found one possible solution, which is wrapping list with np.array. But unfortunately I found that each of the point clouds contains different point number, which makes it impossible to convert this list to np.array.
